scalyr_agent/tests/scalyr_client_test.py

```diff
@@ -905,3 +908,43 @@ def get_user_agent():
         frags = ["frag1", "frag2", "frag3"]
         session.augment_user_agent(frags)
         self.assertEquals(get_user_agent(), base_ua + ";" + ";".join(frags))
+
+    @mock.patch('scalyr_agent.scalyr_client.log')
+    @mock.patch('scalyr_agent.scalyr_client.time.time', mock.Mock(return_value=11111))
+    def test_send_request_body_is_logged_raw_uncompressed(self, mock_log):
+        """
+        When sending a request with compression available / enabled, raw (uncompressed) request
+        body (payload) should be logged under DEBUG log level.
+        """
+        self.assertEqual(mock_log.info.call_count, 0)
+
+        session = ScalyrClientSession(
+            "https://dummserver.com", "DUMMY API KEY", SCALYR_VERSION
+        )
+        # One for "using server" and other for "using session" message
+        self.assertEqual(mock_log.info.call_count, 2)
+        self.assertEqual(mock_log.log.call_count, 0)
```

Comment:
I agree - I'm usually not a big fan of mocking myself (there needs to be a balance) since it's usually testing the implementation details which may change and that can make tests less robust / code more coupled to the tests.

Another approach I could go with is to configure logging to a file and simply check that file contains a specific log line.

In fact, that's what I've seen in a couple of other places so perhaps also a small TestCaseWithLogCapture or similar base class refactor is also a good idea.

Reply:
Here is the approach I end up going with - 7da48802944a23cc5e25a6b459b7597628d9372b, 8086b58b7d514fdf85d10dbbc43f01368719d3a7.

It should be more robust since it doesn't rely on mocking the internals / implementation details. 

I still left one mock check in place though, because I still want to check that body was indeed compressed (aka compression function was called) since that's the edge case that this test is supposed to test.

---

First I wanted to go with a more generic approach where the log output could either be stored to a file or to a class variable (developer would either specify a file handle or StringIO object), but because of the way ``set_log_destination`` function is currently structured, this would either require:

1. Major refactoring of that function
2. "Hacking" approach which relies on ``use_stdout=True`` behavior and mocking ``sys.stdout`` (less code than 1), but on the hack-ish side)

I'm not a huge fan of either of those approaches and I wanted to keep this PR as simple and small as possible, so I went with the "write to files" approach.

As mentioned above, this brings those tests closer to "integration" tests spectrum, but I don't think that's a big deal and in fact, that's already done by various existing unit test cases today.

---

@czerwingithub let me know what you think.

In the future, I would also be OK with implementing either 1) or 2), especially if multiple tests utilize that functionality since it would allow us to avoid disk writes inside the unit tests.

If you are OK with this approach, I can update some of the existing test cases to use that class (but if you think I should leave this out here and keep the scope down, I can also do it in a separate add-on PR once this one is merged or not do it at all at this point).